File: 03_Sliding_Window/03_longest_reapeating_substring/main.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        max_freq = 0
        left = 0
        max_length = 0

        for right in range(len(s)):
            count[s[right]] = count.get(s[right], 0) + 1
            max_freq = max(max_freq, count[s[right]])

            while (right - left + 1) - max_freq > k:
                count[s[left]] -= 1
                left += 1

            max_length = max(max_length, right - left + 1)

        return max_length

# Example usage
solution = Solution()
s1 = "XYYX"
k1 = 2
print(f"Length of longest substring after replacements in '{s1}': {solution.characterReplacement(s1, k1)}")

s2 = "AAABABB"
k2 = 1
print(f"Length of longest substring after replacements in '{s2}': {solution.characterReplacement(s2, k2)}")

File: 03_Sliding_Window/03_longest_reapeating_substring/main.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> count;
        int max_freq = 0;
        int left = 0, max_length = 0;

        for (int right = 0; right < s.size(); ++right) {
            count[s[right]]++;
            max_freq = max(max_freq, count[s[right]]);

            while ((right - left + 1) - max_freq > k) {
                count[s[left]]--;
                left++;
            }

            max_length = max(max_length, right - left + 1);
        }

        return max_length;
    }
};

// Example usage
int main() {
    Solution solution;
    string s1 = "XYYX";
    int k1 = 2;
    cout << "Length of longest substring after replacements in '" << s1 << "': " << solution.characterReplacement(s1, k1) << endl;

    string s2 = "AAABABB";
    int k2 = 1;
    cout << "Length of longest substring after replacements in '" << s2 << "': " << solution.characterReplacement(s2, k2) << endl;

    return 0;
}
